activerecord/lib/active_record/reflection.rb

```diff
@@ -757,6 +759,56 @@ def derive_foreign_key(infer_from_inverse_of: true)
           else
             active_record.model_name.to_s.foreign_key
           end
+
+          if active_record.has_query_constraints?
+            derived_fk = derive_fk_query_constraints(active_record, derived_fk)
+          end
+
+          derived_fk
+        end
+
+        def derive_fk_query_constraints(klass, foreign_key)
+          primary_query_constraints = klass.query_constraints_list
+          owner_pk = klass.primary_key
+
+          if primary_query_constraints.size != 2
+            raise ArgumentError, <<~MSG.squish
+            The query constraints list on the #{klass} model has more than 2
+            attributes. Active Record is unable to derive the query constraints
+            for the association. You must explicitly define the query constraints
+            for this association.
+            MSG
+          end
+
+          if !primary_query_constraints.include?(owner_pk)
+            raise ArgumentError, <<~MSG.squish
+            The query constraints on the #{klass} model does not include the primary
+            key so Active Record is unable to derive the foreign key constraints for
+            the association. You must explicitly define the query constraints for this
+            association.
+            MSG
+          end
+
+          # The primary key and foreign key are both already in the query constraints
+          # so we don't want to derive the key. In this case we want a single key.
+          if primary_query_constraints.include?(owner_pk) && primary_query_constraints.include?(foreign_key)
+            return foreign_key
+          end
+
+          first_key, last_key = primary_query_constraints
+
+          if first_key == owner_pk
+            [foreign_key, last_key.to_s]
+          elsif last_key == owner_pk
+            [first_key.to_s, foreign_key]
+          else
+            raise ArgumentError, <<~MSG.squish
+            Active Record couldn't correctly interpret the query constraints
+            for the #{klass} model. The query constraints on #{klass} are
+            #{primary_query_constraints} and the foreign key is #{foreign_key}.
+            Do you need to explicitly set the query constraints on the association?
+            MSG
```

Comment:
There is no test for this case because it shouldn't happen. But if it does happen the error will give an indication that we can't infer the query constraints and they need to be set explicitly. I could return the foreign key passed in here instead but I'm not sure it would be correct so for now I'd rather raise an error.